### ISSUE TITLE
Serving Trained model for pytorch

### ISSUE THREAD

**Issue by gautamkmr** [pytorch/pytorch#19190]
Tensorflow has Tensorflow Serving. How do people serve models trained with pytorch. 
I see Pytorch has `save` and `load` API. Is there any standard way to server trained model using PyTorch?

**Comment by gautamkmr** [comment]
@soumith  Any suggestion on this ?

**Comment by ezyang** [comment]
cc @jeffreyksmithjr

**Comment by Amiralae** [comment]
It's known that in pytorch, there is no serving for production code level as in tensorflow. 
But you can use ONNX to convert your model to other frameworks that support inference and serving in production. 
One other option might be using kubeflow, find more details here :
https://www.kubeflow.org/docs/components/pytorchserving/
In case you want to use your model locally and test it, then simply use load by specifying the input parameters dictionary of your model, as an example :
    # saving the model on device
    torch.save(model.state_dict(), './name_you_give_to_file.pth')
    #load the model on gpu and test the saved model using gpu
    device = torch.device('cuda') (specify if you want use cpu or gpu, 'cuda' for gpu)
    model1 = your_defined_class_model()  (here you give your class model)
    model1.load_state_dict(torch.load('./name_you_give_to_file.pth')) (load the state dict)
    model1.to(device)  
    model1.eval()  (set at evaluation mode)
Hope it's helpful !

**Comment by Amiralae** [comment]
@gautamkmr  
Better use ONNX to convert your model from pytorch to tensorflow, https://pytorch.org/docs/stable/onnx.html
You trained your model with pytorch and save the parameters with state_dict(). Next you use torch.onnx.export, so you convert your model to ONNX format that you will use to convert from ONNX to Tensorflow then you can serve your model with tensorflow.
A detailed example here : https://github.com/onnx/tutorials/blob/master/tutorials/PytorchTensorflowMnist.ipynb

**Comment by jeffreyksmithjr** [comment]
We run PyTorch in production at Facebook, at enormous scale, so we've actually invested a lot in production serving capabilities and will only continue to improve things. It really just depends on what your use case is.

Running inference on a trained model is as simple as `module(input)`, whether that module is traced using the JIT. The TorchScript JIT is one of the most exciting new features focused on this production use case. It allows for compilation of your high-level Python program. Here's a good example of how to use the hybrid frontend of the JIT: https://pytorch.org/tutorials/beginner/hybrid_frontend/learning_hybrid_frontend_through_example_tutorial.html I also like this tutorial showing you how to take advantage of LibTorch for Python-free C++ serving: https://pytorch.org/tutorials/beginner/saving_loading_models.html 

Of course, you can also take advantage of ONNX export as @Amiralae suggests. This can be useful in combination with mobile runtimes or other runtimes that support ONNX on the server side, like NVIDIA's TensorRT or AWS Elastic Inference.

If you want the model management features of a model server, you can get some of that from the cloud services that run PyTorch and ONNX. There are also several groups trying to build PyTorch compatible OSS model servers. As @Amiralae mentions, there's Kubeflow, which I would characterize as in development. Similarly, [MLFlow](https://mlflow.org/) is also working on a level PyTorch support within their model server. And the are PyTorch community projects, like [EuclidesDB](https://euclidesdb.readthedocs.io/en/latest/): Lots of innovation going on in open source model serving these days.

Hope that helps.

**Comment by parano** [comment]
Hi @gautamkmr - I've been working on a library call BentoML https://github.com/bentoml/BentoML designed to make model serving easy for data scientists,  it has support for PyTorch, and very easy to get started.

Here are the two PyTorch & BentoML example notebooks we have in the repo:
https://colab.research.google.com/github/bentoml/gallery/blob/master/pytorch/fashion-mnist/pytorch-fashion-mnist.ipynb
https://colab.research.google.com/github/bentoml/gallery/blob/master/pytorch/cifar10-image-classification/pytorch-cifar10-image-classification.ipynb

For extreme performance-oriented model serving, I think @Amiralae has pointed out the only option is converting your trained model to ONNX and use with TensorFlow serving. Things like Kubeflow requires you to build an API serving container yourself, which add tones of complexity to the workflow. So we built BentoML to solve exactly this problem.

**Comment by marcindulak** [comment]
You don't need kubeflow to use tensorflow-serving https://www.tensorflow.org/tfx/serving/serving_basic
